Mathlib/SetTheory/Ordinal/Notation/FundamentalSequence.lean

```diff
@@ -0,0 +1,186 @@
+
+@[simp] theorem not_mem_empty (x : α) : x ∉ (∅ : Sequence α) := id
+@[simp] theorem mem_singleton_iff {x y : α} : x ∈ ({y} : Sequence α) ↔ x = y := Iff.rfl
+@[simp] theorem mem_ofFun_iff {x : α} {f : ℕ → α} : x ∈ ofFun f ↔ x ∈ Set.range f := Iff.rfl
+@[simp] theorem mem_range_iff {s : Sequence α} {x : α} : x ∈ s.range ↔ x ∈ s := Iff.rfl
+
+theorem mem_singleton (x : α) : x ∈ ({x} : Sequence α) := mem_singleton_iff.2 rfl
+theorem mem_ofFun {f : ℕ → α} (n : ℕ) : f n ∈ ofFun f := ⟨n, rfl⟩
+
+/-- Maps a sequence through a function -/
+def map (s : Sequence α) (g : α → β) : Sequence β :=
+  match s with
+  | Sum.inl none => ∅
+  | Sum.inl (some x) => {g x}
+  | Sum.inr f => ofFun (g ∘ f)
+
+@[simp] theorem map_empty (g : α → β) : map ∅ g = ∅ := rfl
+@[simp] theorem map_singleton (x : α) (g : α → β) : map {x} g = {g x} := rfl
+@[simp] theorem map_ofFun (f : ℕ → α) (g : α → β) : map (ofFun f) g = ofFun (g ∘ f) := rfl
+
+@[simp]
+theorem map_eq_empty_iff {s : Sequence α} {g : α → β} : s.map g = ∅ ↔ s = ∅ := by
+  apply s.recOn <;> simp
+
+@[simp]
+theorem mem_map {s : Sequence α} {f : α → β} {b : β} : b ∈ s.map f ↔ ∃ a ∈ s, f a = b :=
+  match s with
+  | Sum.inl none => by simp
+  | Sum.inl (some x) => by simp [eq_comm]
+  | Sum.inr g => by simp
+
+/-- Attach to a sequence the proof that it contains all its elements -/
+def attach : (s : Sequence α) → Sequence {a : α // a ∈ s}
+  | Sum.inl none => ∅
+  | Sum.inl (some x) => {⟨x, rfl⟩}
+  | Sum.inr f => ofFun fun n ↦ ⟨f n, n, rfl⟩
+
+@[simp] theorem attach_empty : (∅ : Sequence α).attach = ∅ := rfl
+@[simp] theorem attach_singleton (x : α) : ({x} : Sequence α).attach = {⟨x, rfl⟩} := rfl
+@[simp] theorem attach_ofFun (f : ℕ → α) : (ofFun f).attach = ofFun fun n ↦ ⟨f n, n, rfl⟩ := rfl
+
+@[simp]
+theorem attach_eq_empty_iff {s : Sequence α} : s.attach = ∅ ↔ s = ∅ := by
+  apply s.recOn <;> simp
+
+@[simp]
+theorem mem_attach {s : Sequence α} {x : α} : ∀ h : x ∈ s, ⟨x, h⟩ ∈ s.attach := by
+  apply s.recOn <;> simp
+
+/-- Partial map -/
+def pmap (s : Sequence α) (f : ∀ x ∈ s, β) : Sequence β :=
+  s.attach.map fun x ↦ f x.1 x.2
+
+@[simp]
+theorem pmap_empty (f : ∀ x ∈ (∅ : Sequence α), β) : pmap ∅ f = ∅ :=
+  rfl
+
+@[simp]
+theorem pmap_singleton (y : α) (f : ∀ x ∈ ({y} : Sequence α), β) : pmap _ f = {f y rfl} :=
+  rfl
+
+@[simp]
+theorem pmap_ofFun (g : ℕ → α) (f : ∀ x ∈ ofFun g, β) :
+    pmap _ f = ofFun fun n ↦ f (g n) (Set.mem_range_self _) :=
+  rfl
+
+@[simp]
+theorem pmap_eq_empty_iff {s : Sequence α} : {f : ∀ x ∈ s, β} → pmap _ f = ∅ ↔ s = ∅ := by
+  apply s.recOn <;> simp
+
+@[simp]
+theorem mem_pmap {s : Sequence α} {f : ∀ x ∈ s, β} {b : β} :
+    b ∈ s.pmap f ↔ ∃ (a : α) (h : a ∈ s), f a h = b := by
+  simp [pmap]
+
+/-- Builds a list with the first `n` elements of the sequence. This can be used to print the
+sequence. -/
+def toList (s : Sequence α) (n : ℕ) : List α :=
+  match s with
+  | Sum.inl none => []
+  | Sum.inl (some x) => [x]
+  | Sum.inr f => (List.range n).map f
```

Comment:
This one doesn't have any API for now - let's either give api here or leave it for a future PR which can add that instead (eg that `(toList s n).length <= n`)

Reply:
My intended usage is to call something like `#eval s.toList 10` in order to see the first few terms of the fundamental sequence. So having no API is actually fine.

Note that the lemma you write is false for `n = 0`, as I'm hardcoding `toList {x} = [x]`. Of course, we could do something like `toList {x} = if n = 0 then [] else [x]`, but given that I don't plan on using this definition mathematically, I wouldn't bother with this.

Reply:
> My intended usage is to call something like `#eval s.toList 10` in order to see the first few terms of the fundamental sequence. So having no API is actually fine.

Unfortunately this isn't sufficient justification for not having API for a definition - even if the intended usage is in a particular way, there still should be lemmas describing it. Mathlib isn't just used by you!

Reply:
I'll add the projection lemmas. If someone needs this definition for someone else, then we can figure out if `toList {x}` should be redefined, what other API we should have, and so on.